cpp/src/io/parquet/chunk_dict.cu

```diff
@@ -0,0 +1,372 @@
+/*
+ * Copyright (c) 2021, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <io/parquet/parquet_gpu.hpp>
+
+#include <cudf/detail/iterator.cuh>
+#include <cudf/detail/utilities/cuda.cuh>
+#include <cudf/table/row_operators.cuh>
+
+#include <rmm/exec_policy.hpp>
+
+namespace cudf {
+namespace io {
+namespace parquet {
+namespace gpu {
+
+template <int block_size>
+__global__ void __launch_bounds__(block_size, 1)
+  gpuInitializeChunkHashMaps(device_span<EncColumnChunk> chunks)
+{
+  auto chunk = chunks[blockIdx.x];
+  auto t     = threadIdx.x;
+  // fut: Now that per-chunk dict is same size as ck.num_values, try to not use one block per chunk
+  for (size_t i = 0; i < chunk.dict_map_size; i += block_size) {
+    if (t + i < chunk.dict_map_size) {
+      new (&chunk.dict_map_slots[t + i].first) map_type::atomic_key_type{KEY_SENTINEL};
+      new (&chunk.dict_map_slots[t + i].second) map_type::atomic_mapped_type{VALUE_SENTINEL};
+    }
+  }
+}
+
+template <typename T>
+struct equality_functor {
+  column_device_view const& col;
+  __device__ bool operator()(size_type lhs_idx, size_type rhs_idx)
+  {
+    // We don't call this for nulls so this is fine
+    return equality_compare(col.element<T>(lhs_idx), col.element<T>(rhs_idx));
+  }
+};
```

Comment:
Why don't you just use `element_equality_comparator`? If there is not any nulls, then why not `element_equality_comparator<false>`?

Reply:
Because it requires a second type dispatch. The place we use `equality_functor` has already been type dispatched. Sure, it may be compiled out but it may also not be. That is why element_compare is a separate functor. For when you already know the type.